### PR TITLE
feat(cli): add mode displayEvents

### LINKED ISSUE #36: [CLI] Add a mode which displays events

### PR DESCRIPTION
**Pull request by heafalan**
Fixes #36

### CODE REVIEW

File: packages/cli/main.js [hunk 1]
```diff
@@ -2,23 +2,44 @@
...
 
-  const response = parse(
-    await hrp
-      .post("http://localhost:4000/api/", {
-        body: format.request(0, method, params),
-      })
-      .readAll("utf-8")
-  );
+  if (method === "displayEvents") {
+    let websocket = new WebSocket("ws://localhost:4000");
```
Comment (badrAZ) [inline packages/cli/main.js]: `let` => `const` because it will not be reassigned.

File: packages/cli/main.js [hunk 2]
```diff
@@ -2,23 +2,44 @@
...
+    };
+    websocket.onmessage = event => {
+      const message = parse(event.data);
+      if (message.type !== "notification") {
+        return;
+      }
+      console.log(
+        "method: %s\n %s\n",
+        message.method,
+        util.inspect(message.params, { showHiddent: true })
```
Comment (badrAZ) [inline packages/cli/main.js]: Why not `JSON.stringify` ?

**Review by badrAZ — CHANGES_REQUESTED** [review]

File: packages/cli/main.js [hunk 3]
```diff
@@ -2,23 +2,44 @@
...
   const method = params._[0];
 
-  const response = parse(
-    await hrp
-      .post("http://localhost:4000/api/", {
-        body: format.request(0, method, params),
-      })
-      .readAll("utf-8")
-  );
+  if (method === "displayEvents") {
```
Comment (badrAZ) [inline packages/cli/main.js]: `displayEvents` is not an api method. IMHO, we should use this syntax `cli --watch` to watch events.

Comment (julien-f) [reply packages/cli/main.js]: This could help you: https://github.com/JorgeBucaran/getopts#optsstopearly

File: packages/cli/main.js [hunk 4]
```diff
@@ -2,23 +2,43 @@
...
-      .readAll("utf-8")
-  );
+  if (params.watch) {
+    const websocket = new WebSocket("ws://localhost:4000");
 
-  if (response.type === "response") {
-    console.log(response.result);
-  } else if (response.type === "error") {
-    console.error(response.error);
+    websocket.onerror = function(event) {
```
Comment (badrAZ) [inline packages/cli/main.js]: why not use an arrow function like 'onmessage' ?

File: packages/cli/main.js (hunk 2 above)
Comment (julien-f) [reply packages/cli/main.js]: You did not reply to this question, why did you decide to use `inspect` instead of `JSON.stringify`?
Any particular reason?

Comment (heafalan) [reply packages/cli/main.js]: There were no particular reasons.

File: packages/cli/main.js (hunk 1 above)
Comment (julien-f) [reply packages/cli/main.js]: Please rebase your PR on top of `master`.

File: packages/cli/main.js [hunk 5]
```diff
@@ -4,24 +4,44 @@ import getopts from "getopts";
...
   const params = mapValues(omit(opts, "_"), String);
 
-  const response = parse(
-    await hrp
-      .post("http://localhost:4000/api/", {
-        body: format.request(0, method, params),
-      })
-      .readAll("utf-8")
-  );
+  if (params.watch) {
```
Comment (julien-f) [inline packages/cli/main.js]: What if the method we are calling has a `watch` param?

Comment (heafalan) [reply packages/cli/main.js]: In this case, we don't pay attention to the method. 
Should we execute the method and "watch" after ?

Comment (julien-f) [reply packages/cli/main.js]: No, you should differentiate between options to the main command and options to subcommands (methods).

Check out the `stopEarly` option of `getopts`.

File: packages/cli/main.js [hunk 6]
```diff
@@ -4,19 +4,40 @@ import getopts from "getopts";
 import hrp from "http-request-plus";
 import { format, parse } from "json-rpc-protocol";
 import { mapValues, omit } from "lodash";
-const WebSocket = require("ws");
+import * as WebSocket from "ws";
```
Comment (badrAZ) [inline packages/cli/main.js]: We are using the `ws` default class that's why `import * as ...`  is not necessary. Simply use `import WebSocket from "ws"`.

File: packages/cli/main.js [hunk 7]
```diff
@@ -4,19 +4,40 @@ import getopts from "getopts";
...
+  });
+
+  if (opts.watch) {
     const websocket = new WebSocket("ws://localhost:4000");
 
     websocket.onerror = event => {
       console.error(event);
     };
+
+    if (opts.watch !== true) {
```
Comment (badrAZ) [inline packages/cli/main.js]: `opts.watch` is an option of type `boolean`. It cannot be a name of a method.

File: packages/cli/main.js [hunk 8]
```diff
@@ -4,19 +4,40 @@ import getopts from "getopts";
...
 import { format, parse } from "json-rpc-protocol";
 import { mapValues, omit } from "lodash";
-const WebSocket = require("ws");
+import * as WebSocket from "ws";
 
 async function main() {
-  const opts = getopts(process.argv.slice(2));
-  const method = opts._[0];
-  const params = mapValues(omit(opts, "_"), String);
+  async function call(method, params) {
```
Comment (badrAZ) [inline packages/cli/main.js]: This function is not needed. We only request the server when the watch mode is `undefined`.

Comment (badrAZ) [reply packages/cli/main.js]: As a solution, declare a constant dedicated to the got arguments, call the `getopts` with `stopEarly` as an option then check if the `watch` exist.
If it exists, subscribe to the server to get notifications.
If not, call the `getopts` without `stopEarly` and omit, cast the got params then request the server.

**Review by badrAZ — CHANGES_REQUESTED** [review]

File: packages/cli/main.js (hunk 7 above)
Comment (heafalan) [reply packages/cli/main.js]: Let me describe all the possible cases and their result of opts.watch : 
- For `cli createEntry --name Name1` , `opts.watch` is undefined. 
- For `cli --watch`, sure : `opts.watch=true`.
- But for `cli --watch createEntry --name Name2` (or other method with its options), `opts.watch=createEntry` ( `opts = { _: [], watch: "createEntry", name: "Name2" }` )

File: packages/cli/main.js [hunk 9]
```diff
@@ -1,27 +1,47 @@
...
+        return;
+      }
+      console.log(
+        "method: %s\n %s\n",
+        message.method,
+        JSON.stringify(message.params)
+      );
+    };
+  } else {
+    opts = getopts(opts._);
```
Comment (badrAZ) [inline packages/cli/main.js]: We should not use the result of the first `getopts` call.

**Review by badrAZ — CHANGES_REQUESTED** [review]

File: packages/cli/main.js [hunk 10]
```diff
@@ -1,27 +1,47 @@
...
 import hrp from "http-request-plus";
 import { format, parse } from "json-rpc-protocol";
 import { mapValues, omit } from "lodash";
+import WebSocket from "ws";
 
 async function main() {
-  const opts = getopts(process.argv.slice(2));
-  const method = opts._[0];
-  const params = mapValues(omit(opts, "_"), String);
+  let opts = getopts(process.argv.slice(2), {
```
Comment (badrAZ) [inline packages/cli/main.js]: We can declare a constant for the got args. here is my proposition.
```js
const args = process.argv.slice(2);
const { watch } = getopts(args, { stopEarly: true });
if (watch) { 
} else {
const opts = getopts(args)
}
```

**Review by badrAZ — APPROVED** [review]

File: packages/cli/main.js (hunk 7 above)
Comment (julien-f) [reply packages/cli/main.js]: You can configure `getopts` to always parse `watch` as a boolean.

File: packages/cli/main.js (hunk 8 above)
Comment (julien-f) [reply packages/cli/main.js]: Always run `getopts` with `stopEarly` first to parse global options then call it again without `stopEarly` to parse method parameters.

File: packages/cli/main.js [hunk 11]
```diff
@@ -1,27 +1,46 @@
...
 import { format, parse } from "json-rpc-protocol";
 import { mapValues, omit } from "lodash";
+import WebSocket from "ws";
 
 async function main() {
-  const opts = getopts(process.argv.slice(2));
-  const method = opts._[0];
-  const params = mapValues(omit(opts, "_"), String);
+  const args = process.argv.slice(2);
+  const { watch } = getopts(args, { stopEarly: true });
```
Comment (julien-f) [inline packages/cli/main.js]: Configure `getopts` to always parse `watch` as a boolean.

File: packages/cli/main.js [hunk 12]
```diff
@@ -1,27 +1,49 @@
...
+        return;
+      }
+      console.log(
+        "method: %s\n %s\n",
+        message.method,
+        JSON.stringify(message.params)
+      );
+    };
+  } else {
+    const opts = getopts(args);
```
Comment (julien-f) [inline packages/cli/main.js]: Here you should parse only the options not taken care of by the previous `getopts`:

```js
const { watch, _: optionsLeft } = getopts(/* ... */)
const opts = getopts(/* ... */)
```

File: packages/cli/main.js (hunk 6 above)
Comment (julien-f) [reply packages/cli/main.js]: @badrAZ Those two syntaxes are completely different.

```js
import WebSocket from 'ws' 
// is the same as
import { default as WebSocket } from 'ws'
```

```js
import * as WebSocket from 'ws'

WebSocket.default
```

Comment (julien-f) [reply packages/cli/main.js]: They are only similar due to Babel's behavior when importing a CommonJS module.

File: packages/cli/main.js [hunk 13]
```diff
@@ -1,27 +1,48 @@
-/* eslint-disable no-console */
```
Comment (julien-f) [inline packages/cli/main.js]: Why did you remove this?

**Review by julien-f — APPROVED** [review]
